### PR TITLE
use evoSuite with nopol

### PR DESCRIPTION
**Pull request by dufaux**
I open this pull request to discuss about it.
I have few question:

I don't know where I can put a readme or kind of example to launch the main?

I used a evosuite jar instead of launch evoSuite on java because we got unsolved problem with nopol after evoSuite execution.

Sometime I got java heap space or stack over flow error with nopol on big project. Maybe it is due to my old computer but should I launch nopol in thread and kill it after too long time or not?

I don't know what is the best way to generate multiple jar from the pom.xml, I read that it was not recommended.

### CODE REVIEW

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.List;
+
+import spoon.processing.AbstractProcessor;
+import spoon.reflect.declaration.CtAnnotation;
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtMethod;
+
+/**
+ * This processor remove all method which are not contains in keptMethod and methodTested
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java]: This processor removes all methods that are not  methodTested and not in keptMethods list.

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+ * just compare the signatures of methods.
+ * @author dufaux
+ *
+ */
+public class TestSelectionProcessor extends AbstractProcessor<CtMethod>{
+
+	private CtMethod methodTested;
+	private List<CtMethod> keptMethods;
+
+	public TestSelectionProcessor(CtMethod methodTested, List<CtMethod> keptMethods ){
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java]: You can put methodTested in keptMethods,you don't need two parameters

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+		this.methodTested = methodTested;
+		this.keptMethods = keptMethods;
+	}
+
+	public void process(CtMethod element) {
+		
+		//getEnvironment().setAutoImports(true);
+		boolean isTest = false;
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java]: Best practice You can put line 31-53 in the method 

`````` Java
@Override
public boolean isToBeProcessed(CtMethod element){
    return super.isToBeProcessed(element);
}```
``````

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+		for(CtAnnotation<? extends Annotation> annotation : element.getAnnotations()){
+			if(annotation.getSignature().equals("@org.junit.Test")){
+				isTest = true;
+			}
+		}
+		
+		if(!isTest){
+			//System.out.println("keep '"+element.getSignature()+"'");
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java]: Remove this line

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			return; //keep this one
+		}
+		
+		if(methodTested != null && element.getSignature().equals(methodTested.getSignature())){
+			//System.out.println("keep '"+element.getSignature()+"'");
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java]: Remove this line

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+			return; //keep this one
+		}
+		
+		for(CtMethod method : keptMethods){
+			if(element.getSignature().equals(method.getSignature())){
+				//System.out.println("keep '"+element.getSignature()+"'");
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/TestSelectionProcessor.java]: Remove this line

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/RemoveEvosuiteEffectsProcessor.java [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+			for(CtAnnotation annotation : annotations){
+				if(UnwantedAnnotations.contains(annotation.getActualAnnotation().annotationType().getName())){
+					removed.add(method);
+					break;
+				}
+			}
+		}
+		//System.out.println("####["+element.getSimpleName()+"] Remove annotated method");
+		for(CtMethod method : removed){
+			//System.out.println("    remove "+method.getSimpleName());
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/RemoveEvosuiteEffectsProcessor.java]: Remove this line

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/RemoveEvosuiteEffectsProcessor.java [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+		for(CtMethod method : methods){
+			List<CtAnnotation<? extends Annotation>> annotations = method.getAnnotations();
+			for(CtAnnotation annotation : annotations){
+				if(UnwantedAnnotations.contains(annotation.getActualAnnotation().annotationType().getName())){
+					removed.add(method);
+					break;
+				}
+			}
+		}
+		//System.out.println("####["+element.getSimpleName()+"] Remove annotated method");
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/RemoveEvosuiteEffectsProcessor.java]: Remove this line

File: nopol/src/main/java/fr/inria/lille/spirals/evo/processors/RemoveEvosuiteEffectsProcessor.java [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+
+import spoon.processing.AbstractProcessor;
+import spoon.reflect.declaration.CtAnnotation;
+import spoon.reflect.declaration.CtClass;
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.declaration.CtMethod;
+
+public class RemoveEvosuiteEffectsProcessor extends AbstractProcessor<CtClass>{
+
+	private static final ArrayList<String> UnwantedAnnotations = new ArrayList<String>();
```
Comment (tdurieux) [inline nopol/src/main/java/fr/inria/lille/spirals/evo/processors/RemoveEvosuiteEffectsProcessor.java]: The convention in java is to use camel case, here it's a constant you can also put it in upper-case

**Comment by dufaux** [comment]
I don't understand why the test fail with maven but run fine under eclipse. 
classpath and classname look right and i can't find gzoltar source to search where is the null pointer exception.
Do you have an idea ?

**Comment by tdurieux** [comment]
It is probably due to invalid classpath (class not compiled, invalid jar location, ...)

**Pull request closed without merging**